### ISSUE TITLE
unseen TCR–epitope pairs

### ISSUE THREAD

**Issue by imerelli** [PaccMann/TITAN#20]
Hi, I’m interested in using TITAN to predict binding for new, unseen TCR–epitope pairs, i.e. those not included in the training/validation sets. I would like to ask:

- Is there a simple example or script available to load a trained model and make predictions on custom pairs of TCRβ CDR3 sequences and peptide epitopes (either AA or SMILES)?
	
- The paper shows attention heatmaps to interpret model focus (e.g., Figure 4). I would love to replicate this. Could you provide me thec ode or a notebook to extract and visualize attention weights from a trained model?

This would help a lot for research use cases where interpretability is essential.

Thanks in advance!

**Comment by jannisborn** [comment]
Hi @imerelli ,

(1) Yes a script is available for pure inference: https://github.com/PaccMann/TITAN?tab=readme-ov-file#run-trained-titan-model-on-data

(2) There is no visualization notebook available to do this directly, but if you run above script, you do obtain all information to create such plot. The `model` returns a dictionary, see [here](https://github.com/PaccMann/TITAN/blob/d01eb1effb5ce44c278af5e4b939b62fde4275cc/scripts/flexible_model_eval.py#L226C20-L226C29) which contains various fields, including a confidence score (if you enable it) and also the attention weights on the ligand (here: the epitope) and the receptor (here: the TCR): https://github.com/PaccMann/paccmann_predictor/blob/ed31bd6e74e36525e742709eae01269a069ac6dd/paccmann_predictor/models/bimodal_mca.py#L494
If you extract the values in this Tensor you can create such a heatmap plot like FIg4 in our paper. For 4A we just did that in powerpoint, for 4B we used RDKit.

Let me know if you have more questions, I may be able to point you to the RDKIt code

**Comment by imerelli** [comment]
Ok, thank you for your answer, but how many columns should have name_of_test_data_file.csv ?
Two I guess, since the prediction should be provided by the tool, right? 

Maybe I get confused by this:

test.csv A .csv file containing three columns, one for TCR IDs, one for epitope IDs and one for the labels. This data is used for testing.

**Comment by jannisborn** [comment]
Is this resolved? I dont understand your comment. You're asking about number of columns in the test csv file and below you paste the answer that it has to have three. So you gave the answer already

**Comment by imerelli** [comment]
Sorry for the late response, now that other issues are solved I can focus on this. My question is about the format of the unseen data. I guess it should be something like the test_small.csv, but without the last column, since clearly I don't have the label yet.

Am I right? Something like:

,Unnamed: 0,ligand_name,sequence_id
10,10,28,108
47,47,121,528
48,48,121,533

**Comment by jannisborn** [comment]
I see. You have to mock this column and just impute some made-up labels for the code to run. 
The script was written with the intent to evaluate performance on unseen data where ground truth is known

**Comment by imerelli** [comment]
Ok, I understand, but this is quite uncomfortable to use since in real case I don't have a ground truth. 

I mean, apart from the performance output, where I can get the prediction for each ligand and sequence of your software? This is not clear to me, apart from the issue of specifying a mock column in the input.

By the way, is it possible to have a script that deals with real unseen data? This should take in input a three columns file and provide as results a single column prediction, right?

**Comment by jannisborn** [comment]
Having to edit a single line of code (to mock the labels) is really the opposite of "inconvenient". It takes one minute. This is a 4 year old research project, being you, I would be grateful to receive a response at all. 

Regarding predictions, please read the code, they are saved in the model folder as a `.npy` file.
